components/masking_functions/src/udf/udf_utils_sql.cc

```diff
@@ -0,0 +1,157 @@
+/* Copyright (c) 2023 Percona LLC and/or its affiliates. All rights reserved.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software Foundation,
+   51 Franklin Street, Suite 500, Boston, MA 02110-1335 USA */
+
+#include "components/masking_functions/include/udf/udf_utils_sql.h"
+
+#include <mysql/components/services/mysql_current_thread_reader.h>
+#include <mysql/components/services/security_context.h>
+#include "components/masking_functions/include/component.h"
+#include "components/masking_functions/include/udf/udf_utils.h"
+
+extern REQUIRES_SERVICE_PLACEHOLDER_AS(mysql_command_query, cmd_query_srv);
+extern REQUIRES_SERVICE_PLACEHOLDER_AS(mysql_command_error_info,
+                                       cmd_error_info_srv);
+extern REQUIRES_SERVICE_PLACEHOLDER_AS(mysql_command_query_result,
+                                       cmd_query_result_srv);
+extern REQUIRES_SERVICE_PLACEHOLDER_AS(mysql_command_options, cmd_options_srv);
+extern REQUIRES_SERVICE_PLACEHOLDER_AS(mysql_command_factory, cmd_factory_srv);
+extern REQUIRES_SERVICE_PLACEHOLDER_AS(mysql_thd_security_context,
+                                       thd_security_ctx);
+extern REQUIRES_SERVICE_PLACEHOLDER(mysql_current_thread_reader);
+extern REQUIRES_SERVICE_PLACEHOLDER_AS(mysql_security_context_options,
+                                       security_ctx_options);
+
+extern REQUIRES_SERVICE_PLACEHOLDER(global_grants_check);
+
+constexpr std::string_view privilege_name = "MASKING_DICTIONARIES_ADMIN";
+
+namespace mysql::components {
+
+sql_context::sql_context() {
+  if (cmd_factory_srv->init(&mysql_h)) {
+    throw sql_context_exception("Couldn't initialize server handle");
+  }
+
+  THD *thd;
+  if (mysql_service_mysql_current_thread_reader->get(&thd)) {
+    throw sql_context_exception("Couldn't query current thd");
+  }
+
+  Security_context_handle sctx;
+  if (thd_security_ctx->get(thd, &sctx)) {
+    throw sql_context_exception("Couldn't query security context");
+  }
+
+  if (cmd_options_srv->set(mysql_h, MYSQL_COMMAND_USER_NAME, "root")) {
```

Comment:
Shouldn't we use current user credentials here?

Reply:
No, the documentation states that all users have to be able to use the read only functions, and that the modifications functions only depend on the dynamic privilege created by the component, regardless of access permissions to the internal table.